var/spack/repos/builtin/packages/gromacs/package.py

```diff
@@ -130,9 +130,9 @@ def cmake_args(self):
             options.append('-DGMX_MPI:BOOL=ON')
             # Ensures gmxapi builds properly
             options.extend([
-                '-DMPI_C_COMPILER=%s' % self.spec['mpi'].mpicc,
-                '-DMPI_CXX_COMPILER=%s' % self.spec['mpi'].mpicxx,
-                '-DMPI_Fortran_COMPILER=%s' % self.spec['mpi'].mpifc
+                '-DCMAKE_C_COMPILER=%s' % self.spec['mpi'].mpicc,
```

Comment:
The commit message is at odds with the discussion behind https://gitlab.com/gromacs/gromacs/-/issues/3671
Can you elaborate on what goes wrong? If the documentation at https://manual.gromacs.org/current/install-guide/index.html#mpi-support is wrong, it would be good to get a bug report.

Reply:
Problem was in linking the `gxmapi` portion of the build. The issue manifested in the 2021 release. 2020.5 continues to build successfully without the change. On @2021, I was getting: 'file not recognized: Is a directory' (same errors with g++ v7.3.1, v7.5.0, v8.4.0). 

![image (4)](https://user-images.githubusercontent.com/187202/106926768-0dc5ca00-66d7-11eb-9b22-af360af82bb2.png)

Trial and error found that overriding MPI_*_COMPILER alone was not sufficient. gmxapi linking continues to use g++ for linking.

Reply:
Thanks. Sounds similar to https://gitlab.com/gromacs/gromacs/-/issues/3896 and possibly https://gitlab.com/gromacs/gromacs/-/issues/3882
Can you also provide cmake version?

Reply:
CMake 3.19.2

Full build spec:  
```
[ec2-user@ip-10-0-24-71 spack-0.16]$ spack spec -lI gromacs@2021
Input spec
--------------------------------
 -   gromacs@2021

Concretized
--------------------------------
 -   thiojfq  gromacs@2021%gcc@7.3.1~blas~cuda~cycle_subcounters~double~double_precision+hwloc~ipo~lapack~mdrun_only+mpi~nosuffix~opencl+openmp~plumed+shared~sycl build_type=RelWithDebInfo arch=linux-amzn2-skylake_avx512
[+]  glhmhw6      ^cmake@3.19.2%gcc@7.3.1~doc+ncurses+openssl+ownlibs~qt arch=linux-amzn2-skylake_avx512
[+]  xaetif5          ^ncurses@6.2%gcc@7.3.1~symlinks+termlib arch=linux-amzn2-skylake_avx512
[+]  gkeasfh              ^pkgconf@1.7.3%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  q5wtdsq          ^openssl@1.1.1i%gcc@7.3.1+systemcerts arch=linux-amzn2-skylake_avx512
[+]  ormuls4              ^perl@5.32.1%gcc@7.3.1+cpanm+shared+threads arch=linux-amzn2-skylake_avx512
[+]  urnz3ng                  ^berkeley-db@18.1.40%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  6szhcrm                  ^gdbm@1.18.1%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  a3lxdhq                      ^readline@8.0%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  w4vw6jt              ^zlib@1.2.11%gcc@7.3.1+optimize+pic+shared arch=linux-amzn2-skylake_avx512
[+]  ktkhw65      ^fftw@3.3.9%gcc@7.3.1+mpi~openmp~pfft_patches precision=double,float arch=linux-amzn2-skylake_avx512
[+]  657a6pa          ^openmpi@4.1.0%gcc@7.3.1~atomics~cuda~cxx~cxx_exceptions+gpfs~java+legacylaunchers~lustre~memchecker+pmi~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=ofi schedulers=slurm arch=linux-amzn2-skylake_avx512
[+]  5xqzouq              ^hwloc@2.4.0%gcc@7.3.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-amzn2-skylake_avx512
[+]  pfsqtei                  ^libpciaccess@0.16%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  pkzkm5h                      ^libtool@2.4.6%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  v7fitap                          ^m4@1.4.18%gcc@7.3.1+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-amzn2-skylake_avx512
[+]  tx2zm5l                              ^libsigsegv@2.12%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  cql6yz3                      ^util-macros@1.19.1%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  h6gnrjt                  ^libxml2@2.9.10%gcc@7.3.1~python arch=linux-amzn2-skylake_avx512
[+]  6t7vzh3                      ^libiconv@1.16%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  qhs5lhp                      ^xz@5.2.5%gcc@7.3.1~pic arch=linux-amzn2-skylake_avx512
[+]  toke6jp              ^libevent@2.1.8%gcc@7.3.1+openssl arch=linux-amzn2-skylake_avx512
[+]  a7sea4n              ^libfabric@1.11.0%gcc@7.3.1~kdreg fabrics=efa,mrail,rxd,rxm,shm,sockets,tcp,udp,verbs arch=linux-amzn2-skylake_avx512
[+]  sup7l4c                  ^rdma-core@33.1%gcc@7.3.1~ipo build_type=RelWithDebInfo arch=linux-amzn2-skylake_avx512
[+]  mjv66fg                      ^libnl@3.3.0%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  vipssdp                          ^bison@3.7.4%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  u6ph7eb                              ^diffutils@3.7%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  vyskmfr                              ^help2man@1.47.16%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  3nrdq27                                  ^gettext@0.21%gcc@7.3.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-amzn2-skylake_avx512
[+]  czf7gfa                                      ^bzip2@1.0.8%gcc@7.3.1+shared arch=linux-amzn2-skylake_avx512
[+]  6rup5nv                                      ^tar@1.32%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  fdilp3t                          ^flex@2.6.4%gcc@7.3.1+lex patches=09c22e5c6fef327d3e48eb23f0d610dcd3a35ab9207f12e0f875701c677978d3 arch=linux-amzn2-skylake_avx512
[+]  yt7dp2g                              ^autoconf@2.69%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  7xeeyas                              ^automake@1.16.3%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  3fr36nb                              ^findutils@4.6.0%gcc@7.3.1 patches=84b916c0bf8c51b7e7b28417692f0ad3e7030d1f3c248ba77c42ede5c1c5d11e,bd9e4e5cc280f9753ae14956c4e4aa17fe7a210f55dd6c84aa60b12d106d47a2 arch=linux-amzn2-skylake_avx512
[+]  ixug5h3                                  ^texinfo@6.5%gcc@7.3.1 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-amzn2-skylake_avx512
[+]  4xru65f                      ^py-docutils@0.15.2%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  2cqt6bx                          ^py-setuptools@50.3.2%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  dsgegjc                              ^python@3.8.7%gcc@7.3.1+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-amzn2-skylake_avx512
[+]  mjkj5vh                                  ^expat@2.2.10%gcc@7.3.1+libbsd arch=linux-amzn2-skylake_avx512
[+]  73f5azu                                      ^libbsd@0.10.0%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  p7h3m7b                                  ^libffi@3.3%gcc@7.3.1 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-amzn2-skylake_avx512
[+]  asmyed5                                  ^sqlite@3.34.0%gcc@7.3.1+column_metadata+fts~functions~rtree arch=linux-amzn2-skylake_avx512
[+]  errev24                                  ^util-linux-uuid@2.36%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  evg2vq2              ^numactl@2.0.14%gcc@7.3.1 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94 arch=linux-amzn2-skylake_avx512
[+]  bfznv2e              ^slurm@20-02-4-1%gcc@7.3.1~gtk~hdf5~hwloc~mariadb+pmix+readline~restd sysconfdir=/opt/slurm/etc arch=linux-amzn2-skylake_avx512
[+]  y5vozvl                  ^curl@7.74.0%gcc@7.3.1~darwinssl~gssapi~libssh~libssh2~nghttp2 arch=linux-amzn2-skylake_avx512
[+]  hmn47ro                      ^libidn2@2.3.0%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  ocmfx3a                          ^libunistring@0.9.10%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  2buqgp5                  ^glib@2.66.2%gcc@7.3.1~libmount patches=b3fd45063a19c871048aa1f28692293ab8971a871bdcbe65f06f17fdd79db9e2 tracing=none arch=linux-amzn2-skylake_avx512
[+]  ojazxbv                      ^meson@0.56.0%gcc@7.3.1 patches=7a7ff3bffb5c1996faf178da6765d578394624edb96fe9297bf21f2e3cd9ebda arch=linux-amzn2-skylake_avx512
[+]  qqokxjf                          ^ninja@1.10.2%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  zix2pr4                      ^pcre@8.44%gcc@7.3.1~jit+multibyte+utf arch=linux-amzn2-skylake_avx512
[+]  ehcccnx                  ^json-c@0.13.1%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  jbyafd4                  ^lz4@1.9.2%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  un24hbn                  ^munge@0.5.14%gcc@7.3.1 localstatedir=/var arch=linux-amzn2-skylake_avx512
[+]  6vfj6de                      ^libgcrypt@1.9.1%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  gzsn6aq                          ^libgpg-error@1.37%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  nxo4zlx                              ^gawk@5.1.0%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  ldfyhbc                                  ^gmp@6.1.2%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  3vkib4q                                  ^mpfr@4.0.2%gcc@7.3.1 patches=3f80b836948aa96f8d1cb9cc7f3f55973f19285482a96f9a4e1623d460bcccf0 arch=linux-amzn2-skylake_avx512
[+]  az3knlu                                      ^autoconf-archive@2019.01.06%gcc@7.3.1 arch=linux-amzn2-skylake_avx512
[+]  a6bhf7o                  ^pmix@3.2.1%gcc@7.3.1~docs+pmi_backwards_compatibility~restful arch=linux-amzn2-skylake_avx512
```

Reply:
> gmxapi linking continues to use g++ 

Yes, the reasoning (described in https://gitlab.com/gromacs/gromacs/-/issues/3671) is that it is preferable to try to determine the flags to provide to the wrapped compiler, rather than to use the compiler wrapper in the generated build system. One potential issue is that, if a compiler wrapper is provided, the infrastructure may not adequately check that it is wrapping the same compiler that is determined for CMAKE_CXX_COMPILER. But this seems like something else. Hopefully, the parent commit here gives enough information to reproduce the issue! Thank you Spack! :-)